src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubDispatcher.cs

```diff
@@ -393,6 +396,9 @@ private bool TryGetStreamingEnumerator(HubConnectionContext connection, string i
         {
             if (result != null)
             {
+                // TODO: cache reflection for performance, on HubMethodDescriptor maybe?
+                resultType = UnwrapTask(resultType);
```

Comment:
I spent a few minutes wondering if it was really this simple and thinking you forgot to `await` the result to get the `IObservable` before realizing that you **do** already have the unwrapped `IObservable` in `result`, it's just that `resultType` is the defined return type obtained via reflection rather than the actual type of `result`. That makes me wonder if we should just use `result.GetType()` for all the detection logic instead of `resultType`.

Reply:
What should happen if there is a hub method that is typed to return `object`, and at runtime logic sometimes returns `List<string>` and other times returns `IObservable<string>`. Should it be streamable depending on what is returned?

By basing logic on the methods type definition you are able to cache exactly what you expect.

Reply:
> Should it be streamable depending on what is returned?

Never mind, `ValidateInvocationMode` already enforces based on the method's type definition.

IMO for perf type info should be cached for `ValidateInvocationMode` and we might as well reuse the cache type info when resolving the streaming enumerator.

Reply:
> What should happen if there is a hub method that is typed to return object, and at runtime logic sometimes returns List<string> and other times returns IObservable<string>

Yeah, that was my concern too. I suppose this is still a better way to do this.

Reply:
Yea this is fine. I think mvc uses the runtime type but this is a bit cleaner